bundle/libraries/upload.go

```diff
@@ -0,0 +1,268 @@
+
+type configLocation struct {
+	configPath dyn.Path
+	location   dyn.Location
+}
+
+type libsLocations = map[string]([]configLocation)
+
+// Collect all libraries from the bundle configuration and their config paths.
+// By this stage all glob references are expanded and we have a list of all libraries that need to be uploaded.
+// We collect them from task libraries, foreach task libraries  environment dependencies and artifacts.
+// We return a map of library source to a list of config paths and locations where the library is used.
+// We use map so we don't upload the same library multiple times.
+// Instead we upload it once and update all the config paths to point to the uploaded location.
+func collectLocalLibraries(b *bundle.Bundle) libsLocations {
+	libs := make(map[string]([]configLocation))
+
+	for i, job := range b.Config.Resources.Jobs {
+		for j, task := range job.Tasks {
+			for k, lib := range task.Libraries {
+				if !IsLocalLibrary(&lib) {
+					continue
+				}
+
+				p := dyn.NewPath(
+					dyn.Key("resources"),
+					dyn.Key("jobs"),
+					dyn.Key(i),
+					dyn.Key("tasks"),
+					dyn.Index(j),
+					dyn.Key("libraries"),
+					dyn.Index(k),
+				)
+				if lib.Whl != "" {
+					libs[lib.Whl] = append(libs[lib.Whl], configLocation{
+						configPath: p.Append(dyn.Key("whl")),
+						location:   b.Config.GetLocation(p.String()),
+					})
+				}
+
+				if lib.Jar != "" {
+					libs[lib.Jar] = append(libs[lib.Jar], configLocation{
+						configPath: p.Append(dyn.Key("jar")),
+						location:   b.Config.GetLocation(p.String()),
+					})
+				}
+			}
+
+			if task.ForEachTask != nil {
+				for l, lib := range task.ForEachTask.Task.Libraries {
+					if !IsLocalLibrary(&lib) {
+						continue
+					}
+
+					p := dyn.NewPath(
+						dyn.Key("resources"),
+						dyn.Key("jobs"),
+						dyn.Key(i),
+						dyn.Key("tasks"),
+						dyn.Index(j),
+						dyn.Key("for_each_task"),
+						dyn.Key("task"),
+						dyn.Key("libraries"),
+						dyn.Index(l),
+					)
+
+					if lib.Whl != "" {
+						libs[lib.Whl] = append(libs[lib.Whl], configLocation{
+							configPath: p.Append(dyn.Key("whl")),
+							location:   b.Config.GetLocation(p.String()),
+						})
+					}
+
+					if lib.Jar != "" {
+						libs[lib.Jar] = append(libs[lib.Jar], configLocation{
+							configPath: p.Append(dyn.Key("jar")),
+							location:   b.Config.GetLocation(p.String()),
+						})
+					}
+				}
+			}
+		}
+
+		for j, env := range job.Environments {
+			if env.Spec == nil {
+				continue
+			}
+
+			for k, dep := range env.Spec.Dependencies {
+				if !IsLibraryLocal(dep) {
+					continue
+				}
+
+				p := dyn.NewPath(
+					dyn.Key("resources"),
+					dyn.Key("jobs"),
+					dyn.Key(i),
+					dyn.Key("environments"),
+					dyn.Index(j),
+					dyn.Key("spec"),
+					dyn.Key("dependencies"),
+					dyn.Index(k),
+				)
+
+				libs[dep] = append(libs[dep], configLocation{
+					configPath: p,
+					location:   b.Config.GetLocation(p.String()),
+				})
+			}
+		}
+	}
```

Comment:
These loops use both the static config and then synthesize the dynamic path. Could we just use `dyn.MapByPattern` to go over all of them? This might also allow for reuse of the patterns we use between the expand glob mutator and this one.